scripts/trim-low-abund.py

```diff
@@ -0,0 +1,295 @@
+
+        screed_iter = screed.open(filename)
+        pass2fp = open(pass2filename, 'w')
+        trimfp = open(trimfilename, 'w')
+
+        save_pass2 = 0
+        for n, is_pair, read1, read2 in broken_paired_reader(screed_iter):
+            if n % 10000 == 0:
+                print >>sys.stderr, '...', n, filename, save_pass2, \
+                    n_reads, n_bp, written_reads, written_bp
+
+            # we want to track paired reads here, to make sure that pairs
+            # are not split between first pass and second pass.
+
+            if is_pair:
+                n_reads += 2
+                n_bp += len(read1.sequence) + len(read2.sequence)
+
+                seq1 = read1.sequence.replace('N', 'A')
+                seq2 = read2.sequence.replace('N', 'A')
+
+                med1, _, _ = ct.get_median_count(seq1)
+                med2, _, _ = ct.get_median_count(seq2)
+
+                if med1 < NORMALIZE_LIMIT or med2 < NORMALIZE_LIMIT:
+                    ct.consume(seq1)
+                    ct.consume(seq2)
+                    write_record_pair(read1, read2, pass2fp)
+                    save_pass2 += 2
+                else:
+                    _, trim_at1 = ct.trim_on_abundance(seq1, CUTOFF)
+                    _, trim_at2 = ct.trim_on_abundance(seq2, CUTOFF)
+
+                    if trim_at1 >= K:
+                        read1 = trim_record(read1, trim_at1)
+
+                    if trim_at2 >= K:
+                        read2 = trim_record(read2, trim_at2)
+
+                    if trim_at1 != len(seq1):
+                        trimmed_reads += 1
+                    if trim_at2 != len(seq2):
+                        trimmed_reads += 1
+
+                    write_record_pair(read1, read2, trimfp)
+                    written_reads += 2
+                    written_bp += trim_at1 + trim_at2
+            else:
+                n_reads += 1
+                n_bp += len(read1.sequence)
+
+                seq = read1.sequence.replace('N', 'A')
+                med, _, _ = ct.get_median_count(seq)
+
+                # has this portion of the graph saturated? if not,
+                # consume & save => pass2.
+                if med < NORMALIZE_LIMIT:
+                    ct.consume(seq)
+                    write_record(read1, pass2fp)
+                    save_pass2 += 1
+                else:                       # trim!!
+                    _, trim_at = ct.trim_on_abundance(seq, CUTOFF)
+                    if trim_at >= K:
+                        new_read = trim_record(read1, trim_at)
+                        write_record(new_read, trimfp)
+
+                        written_reads += 1
+                        written_bp += trim_at
+
+                        if trim_at != len(read1.sequence):
+                            trimmed_reads += 1
+
+        pass2fp.close()
+        trimfp.close()
+
+        print '%s: kept aside %d of %d from first pass, in %s' % \
```

Comment:
On Wed, Feb 11, 2015 at 08:46:29AM -0800, Michael R. Crusoe wrote:
> > +                else:                       # trim!!
> > +                    _, trim_at = ct.trim_on_abundance(seq, CUTOFF)
> > +                    if trim_at >= K:
> > +                        new_read = trim_record(read1, trim_at)
> > +                        write_record(new_read, trimfp)
> > +
> > +                        written_reads += 1
> > +                        written_bp += trim_at
> > +
> > +                        if trim_at != len(read1.sequence):
> > +                            trimmed_reads += 1
> > +
> > +        pass2fp.close()
> > +        trimfp.close()
> > +
> > +        print '%s: kept aside %d of %d from first pass, in %s' % \
> 
> to STDERR

No, this is useful information that I want logged/saved; it's not available
anywhere else, and it's not purely diagnostic.